File: BAEKJOON/2636.py
[200~# 1. 빈칸을 돌면서 빈 칸인지 내부 공간인지 확인한다.
        # 2. 치즈를 돌면서 얼마나 접촉했는지 확인한다.

import sys
input = sys.stdin.readline
def check_air(r:int,c:int,v:set[int]):
    v.add((r,c))
    visited[r][c],flag = True, True
    for k in range(4):
        if not flag:
            break
        row, col = r+dx[k], c+dy[k]
        if 0<= row < N and 0<= col < M:
            if (row,col) not in v: # 방문 안 함
                if not matrix[row][col] and not visited[row][col]: # 갈 수 있는 곳
                    flag = check_air(row,col,v)
                elif visited[row][col]: # visited임
                    flag = False
        else: # 범위를 벗어남
            flag = False
    return flag
def check() -> None:
    output = []
    for i in range(N):
        for j in range(M):
            if matrix[i][j]:
                cnt = 0
                for k in range(4):
                    x, y = dx[k] + i ,dy[k] + j
                    if 0<= x <N and 0<= y <M and not air[x][y]:
                        cnt += 1
                if 1 < cnt:
                    output.append((i,j))
    for i, j in output:
        matrix[i][j] = 0
    return True if output else False

N,M = map(int, input().split())
matrix = [list(map(int, input().split()))for _ in range(N)]
cheeze = True
time,dx,dy = 0,[0,1,0,-1], [1,0,-1,0]
while cheeze:
    visited = [[False for _ in range(M)] for _ in range(N)]
    air = [[matrix[i][j] for j in range(M)] for i in range(N)]
    for i in range(N): # 치즈 내부 공간 찾기
        for j in range(M):
            if not visited[i][j] and not matrix[i][j]:
                v  = set()
                if check_air(i,j,v):
                    for r,c in v:
                        air[r][c] = -1
    cheeze = check()
    time += 1
print(time - 1)

File: BAEKJOON/2636.java
import java.io.BufferedReader;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;
public class Main {
	static int[] dx = {0,1,0,-1};
	static int[] dy = {1,0,-1,0};
 	public static void main (String[] args) throws IOException {
 		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
 		
		String[] input = br.readLine().split(" ");
		char[][] m = new char[Integer.parseInt(input[0])][Integer.parseInt(input[1])];
		for(int k = 0; k < Integer.parseInt(input[0]); k++) {
			m[k] = br.readLine().replaceAll(" ", "").toCharArray();
		}
		int cnt = 0;
		int sol = 0;
		int prev = 0;
		Queue<int []> q = new LinkedList<>();
		while(true) {
			air(m);
			for(int i = 0; i < m.length; i ++ ) {
				for(int j = 0; j < m[0].length; j++) {
					if(m[i][j] == '1' ) {
						cnt ++;
						if(check(i,j,m)) {
							q.add(new int[] {i,j});
						}
					}
				}
			}
			if(cnt == 0) break;
			sol ++;
			prev = cnt;
			cnt = 0;
			while (!q.isEmpty()) {
				int[] node = q.remove();
				m[node[0]][node[1]] = '2';
			}
		}
			System.out.println(sol);
			System.out.println(prev);
 	}
		
 	public static boolean check(int r, int c, char[][] m) {
 		for(int i = 0; i < 4; i ++) {
 			int x = r + dx[i];
 			int y = c + dy[i];
 			if(0<=x && x< m.length && 0<=y  && y < m[0].length && m[x][y] == '2') return true;
 		}
 		return false; // 녹는거 아님
 	}
 	
 	public static void air(char [][] m) {
 		Queue<int []> q = new LinkedList<>();
 		boolean[][] v = new boolean[m.length][m[0].length]; 
 		q.add(new int[] {0,0});
 		m[0][0] = '2';
 		v[0][0]= true;
 		while(!q.isEmpty()) {
 			int[] node = q.remove();
 			for(int i = 0; i < 4; i ++) {
 	 			int x = node[0] + dx[i];
 	 			int y = node[1] + dy[i];
 	 			if(0<=x && x< m.length && 0<=y  && y < m[0].length && !v[x][y] &&(m[x][y] == '0' || m[x][y] == '2' )) {
 	 				m[x][y] = '2';
 	 				q.add(new int[] {x,y});
 	 				v[x][y]=true;
 	 			}
 	 		}
 		}
 	}
 	
}


